bite-size-notes/week-1-e-linearization.md

```diff
@@ -0,0 +1,250 @@
+The code is as follows:
+```python
+# JAX
+import jax.numpy as jnp
+from jax import grad, jacfwd
+
+def pendulum_dynamics(x, omega=1.0, r=0.1):
+    x1, x2 = x
+    dx1dt = x2
+    dx2dt = -omega**2 * jnp.sin(x1) - r * x2
+    return jnp.array([dx1dt, dx2dt])
+
+def linearize_system(x, omega=1.0, r=0.1): 
+    jacobian_func = jacfwd(pendulum_dynamics, argnums=0)
+    jacobian_at_x = jacobian_func(x, omega, r)   
+    return jacobian_at_x
+
+x_example = jnp.array([0.0, 0.0])  # Equilibrium point
+omega = 1.0  # Natural frequency
+r = 0.1      # Damping coefficient
+jacobian_at_equilibrium = linearize_system(x_example, omega, r)
+print("Jacobian matrix at the equilibrium point:", jacobian_at_equilibrium)
+```
+- Other **Snippet codes** you can use for **linearization**:
+```  
+# TensorFlow
+import tensorflow as tf
+@tf.function
+def pendulum_tf(x):
+    x1, x2 = x[0], x[1]
+    return tf.stack([x2, -omega**2 * tf.sin(x1) - r * x2])
+x_tf = tf.Variable([0.0, 0.0])
+with tf.GradientTape() as tape:
+    tape.watch(x_tf)
+    y_tf = pendulum_tf(x_tf)
+A_tf = tape.jacobian(y_tf, x_tf)
+
+# PyTorch
+import torch
+def pendulum_torch(x):
+    x1, x2 = x.unbind()
+    return torch.stack([x2, -omega**2 * torch.sin(x1) - r * x2])
+x_torch = torch.tensor([0.0, 0.0], requires_grad=True)
+y_torch = pendulum_torch(x_torch)
+A_torch = torch.autograd.functional.jacobian(pendulum_torch, x_torch)
+
+# Auto_diff (Pseudocode)
+def pendulum_auto_diff(x, omega=1.0, r=0.1):
+    x1, x2 = x
+    return [x2, -omega**2 * auto_diff.sin(x1) - r*x2]
+x0_auto_diff = [0.0, 0.0]
+A_auto_diff = auto_diff.compute_jacobian(pendulum_auto_diff, x0_auto_diff)
+```
+**TensorFlow** excels in scalable deep learning, offering (tf.GradientTape) for eager automatic differentiation and optimizations via (tf.function). It uniquely integrates with tools like TensorBoard for visualization and TensorFlow Lite for mobile, standing out for deployment capabilities.
+
+**PyTorch** is useful for research with its dynamic computation graphs and intuitive, Pythonic syntax. It's particularly user-friendly for rapid prototyping and iteration, differentiating itself with a more accessible approach to graph construction and debugging.
+
+**Auto_diff (Pseudocode)** conceptualizes automatic differentiation, simplifying the understanding of gradient and Jacobian computation without tying to a specific API(Application Programming Interface), highlighting the core principles of auto differentiation technologies.
+
+## Conclusion
+Linearization simplifies complex systems into easier linear models, crucial for better understanding and control in many areas. This guide shows how to use linearization on a pendulum, with clear math explanations and code examples in JAX, TensorFlow, and PyTorch, showcasing how these tools make analyzing systems straightforward.
```

Comment:
Since you mentioned how linearization make analyzing systems straightforward, somewhere in the Main Body you might want to relate back to linear VS non-linear systems, and show why doing linearization makes it easier for us to, say, analyze the Pendulum.

Reply:
Good idea! Thank you!